src/base/bittorrent/torrentparamrules.h

```diff
@@ -0,0 +1,130 @@
+// category, tags) based on a set of rules related to the torrent's metadata.
+//
+// Rules consist of a set of conditions and a set of modifications. If the
+// conditions are satisfied, the modifications are applied. If multiple rules
+// match a given torrent, modifications are applied in sequential order.
+//
+// Metadata may or may not be available when initially adding the torrent, and
+// there are two API variants to address each of these cases. The first operates
+// on a TorrentDescriptor and AddTorrentParams, and is used when metadata is
+// available prior to adding a torrent to the session. The second operates on a
+// Torrent object, and is used when metadata arrives after adding the torrent.
+
+namespace BitTorrent
+{
+    class TorrentParamRuleComponent : public QObject
+    {
+    public:
+        using QObject::QObject;
+
+        virtual QJsonValue toJson() const = 0;
+    };
+
+    class TorrentParamCondition : public TorrentParamRuleComponent
+    {
+    public:
+        static TorrentParamCondition *fromJson(const QJsonObject &json, QObject *parent);
+
+        virtual ~TorrentParamCondition() = default;
+
+        virtual bool isSatisfied(const TorrentDescriptor &descriptor) const = 0;
+        virtual bool isSatisfied(const Torrent &torrent) const = 0;
+
+    protected:
+        using TorrentParamRuleComponent::TorrentParamRuleComponent;
+    };
+
+    class TorrentParamModifier : public TorrentParamRuleComponent
+    {
+    public:
+        static TorrentParamModifier *fromJson(const QJsonObject &json, QObject *parent);
+
+        virtual ~TorrentParamModifier() = default;
+
+        virtual void modify(AddTorrentParams *params) const = 0;
+        virtual void modify(Torrent *torrent) const = 0;
+
+    protected:
+        using TorrentParamRuleComponent::TorrentParamRuleComponent;
+    };
+
+    class TorrentParamRule : public TorrentParamRuleComponent
+    {
+    public:
+        using TorrentParamRuleComponent::TorrentParamRuleComponent;
+
+        static TorrentParamRule *fromJson(const QJsonObject &json, QObject *parent);
+
+        TorrentParamRule(TorrentParamCondition *condition, TorrentParamModifier *modifier, QObject *parent);
+
+        void apply(const TorrentDescriptor &descriptor, AddTorrentParams *params) const;
+        void apply(Torrent *torrent) const;
+
+        QJsonValue toJson() const override;
+
+    private:
+        TorrentParamCondition *m_condition;
+        TorrentParamModifier *m_modifier;
+    };
+
+    class TorrentParamRules : public QObject
+    {
+    public:
+        using QObject::QObject;
+
+        void addRule(TorrentParamRule *rule);
+
+        void clearRules();
+
+        void apply(const TorrentDescriptor &descriptor, AddTorrentParams *params) const;
+        void apply(Torrent *torrent) const;
+
+        QJsonObject toJson() const;
+        size_t loadRulesFromJson(const QJsonObject &json);
+
+    private:
+        std::vector<TorrentParamRule *> m_rules;
```

Comment:
It is preferable to use Qt types, e.g. QList in this case.

Reply:
Done